/send command; ability to add bots to group chats

alwx opened the pull request:
Done.
Includes several important fixes for `parameter-boxes` and other components.

rasom commented on src/status_im/chat/handlers/input.cljs:
```diff
@@ -332,3 +337,47 @@
+    (fn [{:keys [current-chat-id] :as db} [_ selection]]
+      (let [input-text (get-in db [:chats current-chat-id :input-text])
+            command    (input-model/selected-chat-command db current-chat-id input-text)]
+        (when (and (= selection (+ 2 (count (get-in command [:command :name]))))
```
@alwx it's the code is not super clear here, could you add some comment? I mean this `+ 2` needs some explanation, `(when what-exactly? ...)`

rasom commented on src/status_im/chat/handlers/input.cljs:
```diff
@@ -332,3 +337,47 @@
+  (handlers/side-effect!
+    (fn [{:keys [chat-ui-props current-chat-id] :as db} _]
+      (let [arg-pos (input-model/argument-position db current-chat-id)]
+        (when (> arg-pos 0)
```
`pos?`

rasom commented on src/status_im/chat/handlers/input.cljs:
```diff
@@ -332,3 +337,47 @@
+        (when (> arg-pos 0)
+          (let [input-text (get-in db [:chats current-chat-id :input-text])
+                new-sel    (->> (input-model/split-command-args input-text)
+                                (take (+ 1 arg-pos))
```
`inc`

rasom commented on src/status_im/chat/handlers/input.cljs:
```diff
@@ -332,3 +337,47 @@
+        (let [input-text   (get-in db [:chats current-chat-id :input-text])
+              command-args (cond-> (input-model/split-command-args input-text)
+                                   (input-model/text-ends-with-space? input-text) (conj ""))
+              _ (log/debug "ALWX command-args" input-text command-args arg-pos)
```
debug message

rasom commented on src/status_im/chat/handlers/input.cljs:
```diff
@@ -332,3 +337,47 @@
+                                   (input-model/text-ends-with-space? input-text) (conj ""))
+              _ (log/debug "ALWX command-args" input-text command-args arg-pos)
+              new-sel      (->> command-args
+                                (take (+ 3 arg-pos))
```
Why `+ 3`?

rasom commented on src/status_im/chat/models/input.cljs:
```diff
@@ -81,26 +81,28 @@
+            (if ends-with-space?
+              argument-index
+              (dec argument-index)))
+          -1)))
```
Magic value `-1`, better to have some constant here.

flexsurfer commented:
![image](https://cloud.githubusercontent.com/assets/11790366/26409223/3b896b60-40a0-11e7-9988-bea4ab856c42.png)

alwx commented:
It's finally done.
It was problematic to make commands work in group chats, but it is also implemented. However, there are things that can be implemented, but it's better to postpone them — for instance, if we want to execute `/send` command in 1-to-1 chat without any contact selection, we need to modify the API. I see two options:
1) We can introduce scopes and allow commands to be `group-only` or `1-to-1-only`. By doing this, we will allow users to have different commands with the same name;
2) We can transform command's `parameters` to be functions, and there we will be able to decide how many parameters to return.

Both changes are breaking, and it's worth doing something only after the hackathon.

jarradh commented:
nice work and agreed on both points - they can wait til after hackathon

annadanchenko commented:
@alwx Can't send 1.1234 ETH - no Confirm transaction screen is shown

in 1-1 chat send 1.12345 ETH -> no screen to confirm transaction is shown, app freezes. iPhone 7 plus, iOS 10.3.1
[nosend.log.txt](https://github.com/status-im/status-react/files/1038971/nosend.log.txt)
https://drive.google.com/open?id=0Bz3t9zSg1wb7Wno0Wnllc1pEWjQ

annadanchenko commented:
On Android: password input shows plain text instead of dots:
![screenshot_20170530-173300](https://cloud.githubusercontent.com/assets/6584810/26594079/029d3150-456f-11e7-935e-8252305417d4.png)

annadanchenko commented:
on both iOS and Android: can send /request with not-number, e.g `/request abc`
it's send as Requesting NaN ETH. 
![img_0101](https://cloud.githubusercontent.com/assets/6584810/26594741/80159936-4571-11e7-8cdc-33e0a99f44dd.jpeg)

annadanchenko commented:
both iOS and Android: can `/send non_existing_contact 123` and get error on command handling. for this use /send comand but do not select contact from the list, just type some random word and sum and tap send icon
![screenshot_20170530-221912](https://cloud.githubusercontent.com/assets/6584810/26603497/b74e9c9a-458f-11e7-861f-baac444e3bfd.png)
![screenshot_20170530-221907](https://cloud.githubusercontent.com/assets/6584810/26603496/b74b03dc-458f-11e7-9c99-3d9e29274c08.png)

annadanchenko commented:
if recipient name contains double quote " then it's not possible to send eth:
1. name has " at the end: send here on ios" -> name is shown again for selection after sum is entered. tap on send icon does nothing
![screenshot_20170530-224215](https://cloud.githubusercontent.com/assets/6584810/26604302/9ac71838-4592-11e7-8417-162164943cb7.png)


2. name has " inside it and at the end: send "here on ios" -> no screen Send transaction 
![screenshot_20170530-224247](https://cloud.githubusercontent.com/assets/6584810/26604370/d01820e0-4592-11e7-952b-07721b67845c.png)

alwx commented:
Everything (except https://github.com/status-im/status-react/pull/1227#issuecomment-304930134, because it has been moved to another task) has been fixed. 
@annadanchenko, please, check it.

flexsurfer commented:
On Android: password input shows plain text instead of dots: - still can be reproduced

alwx commented:
Sorry, I've committed the fix.

alwx commented:
So now it should be checked and probably merged :)

annadanchenko commented:
/send Jarrad aaa -> Error on command handling
![img_1553](https://cloud.githubusercontent.com/assets/6584810/26720195/f737e684-478f-11e7-88a8-7d29d4cf56fe.jpeg)




on last build Branch: feature/send-command
Android: http://artifacts.status.im:8081/artifactory/pull-requests/im.status.ethereum-a33832.apk
iOS: https://i.diawi.com/Ytb2EB

annadanchenko commented:
in old `/send` if it’s not enough ether then we show error “Insufficient funds…” when user tap on “send” and do not show transaction to sign. In new `/send`, we make user to confirm transaction and only than show error on Insufficient funds. Expected: works as in old `/send

alwx commented:
I've fixed both issues.

flexsurfer commented on src/status_im/commands/handlers/loading.cljs:
```diff
@@ -16,26 +16,22 @@
+                      (->> (get-in chats [identity :contacts])
+                           (map :identity)
+                           (into [])))]
+    (when (seq contacts)
```
what does this check do?

alwx replied:
what do you mean?
```
cljs.user=> (seq nil)
nil
cljs.user=> (seq [])
nil
cljs.user=> (seq [1 2 3])
(1 2 3)
```

rasom replied:
https://github.com/bbatsov/clojure-style-guide#nil-punning

flexsurfer replied:
thank you, It's not obvious without this link

annadanchenko commented:
In Group chat: if contact A sends to B some ether then there are 2 messages shown:
1. Contact A, empty line, sum
2. Contact A, empty line, NaN ETH
I would expect 1 message with info about who sends ether; to whom ether is sent, sum sent
![screenshot_20170602-130958](https://cloud.githubusercontent.com/assets/6584810/26723833/0da91a78-479f-11e7-8735-baa16daa269d.png)

annadanchenko commented:
In 1-1 chat, Contact A requesting 0.32 ETH. Contact B taps on the request message in chat but input field contains "/send 0.32 09ef186e17b87bc420fe84f4d74d327dpa848af11" and if try to send it then error "Amount is not a valid number" shown. 
"09ef186e17b87bc420fe84f4d74d327dpa848af11" is an address of Contact B that is strange as ether is to be sent to Contact A.
Expected: when user taps on request message then all data is filled in, so it's ok to use Send
![img_1554](https://cloud.githubusercontent.com/assets/6584810/26724799/f5ef8bec-47a3-11e7-93b4-c9f458446686.jpeg)

alwx commented:
@annadanchenko done. 
I've updated both /request and /send commands, checked the communication between two devices (Android and iOS) and updated the code.

It can be tested now.

rasom commented:
<img width=320 src=https://user-images.githubusercontent.com/2364994/27630020-eb59b6b8-5bfc-11e7-946d-0723ccfa0234.PNG />

That's how `/send` and `/receive` are shown in group chat. 
Both send/request were sent to Jarrad's account.

rasom commented:
@alwx please resolve conflicts

annadanchenko commented:
@alwx "Recipient address must be specified" error if tap on Request and try to send valid command. Issue happens only if in the current session in 1-1 chat there were no /send commands send before this. 
video: https://drive.google.com/open?id=0Bz3t9zSg1wb7T1ZuU3BsYWx5X0E
Steps: 
1. Launch app
2. Receive request for 0.123 ETH
3. Tap on request. As a result command `/send "iphone6 send" 0.123` is shown
4. Tap on Send button

Currently: error "Recipient address must be specified" is shown
Expected: Transaction screen to confirm transaction

annadanchenko commented:
@alwx `[missing {{amount}} value] ETH` is shown in group chat to all members (except sender)
![screenshot_20170628-134148](https://user-images.githubusercontent.com/6584810/27633584-79052834-5c08-11e7-86a4-52c083ef4515.png)



1. create group chat
2. As contact A send to contact B 0.23 ETH
3. As contact B open group chat and check messages

Expected: message like X ETH is sent from A to B
Currently: there are 2 messages shown to recipient and other group members:
1. Contact A 0.23 ETH
2. [missing {{amount}} value] ETH - this message is wrong

flexsurfer commented on bots/wallet/bot.js:
```diff
@@ -1,13 +1,280 @@
+    }
+);
+
+function getFeeExplaination(n) {
```
Explanation

flexsurfer commented on bots/wallet/bot.js:
```diff
@@ -1,13 +1,280 @@
+                    status.components.touchable(
+                        {
+                            onPress: status.components.dispatch([status.events.FOCUS_INPUT, []])
+                        },
```
difficult to read code with styles maybe better to move styles in the separate file like in clojure we do?

flexsurfer commented on src/status_im/chat/views/choosers/choose_contact.cljs:
```diff
@@ -0,0 +1,50 @@
+      [contact-view {:contact  contact
+                     :on-press #(select-contact arg-index bot-db-key contact)}])))
+
+(defn choose-contact-view [{arg-index :index}]
```
just interesting why don't you use defview macro, not only here

flexsurfer commented on src/status_im/chat/handlers/commands.cljs:
```diff
@@ -6,15 +6,27 @@
-        (if-not (get-in contacts [jail-id :commands-loaded])
+             :keys [message-id from chat-id on-requested jail-id] :as message} data-type]]
+      (let [jail-id (or jail-id chat-id)
+            jail-id (if (get-in chats [jail-id :group-chat])
```
i saw in code using of `'` for redefine, i think it's good practice `jail-id'`

rasom commented:
1. create acc using `develop` branch
2. use `/sent`/`/receive` commands in 1-1 chat
3. upgrade app using this branch
4. open 1-1 chat with `/send`/`/receive` messages
5. see crash

```
Jul  7 20:02:27 StatusIm[3887] <Notice>: Error rendering component (in Uz > reagent1 > Vz > reagent38 > Lz > reagent45 > Pf > ListView >  > bz > reagent46 > n > Mf > message-content > wz > az > nz)
Jul  7 20:02:27 StatusIm[3887] <Notice>: null is not an object (evaluating 'l.a')
Jul  7 20:02:27 StatusIm[3887] <Notice>: Unhandled JS Exception: null is not an object (evaluating 'l.a')
```

rasom commented:
1. That's how receive/send messages look at first run
<img width=320 src=https://user-images.githubusercontent.com/2364994/27968827-4f3b3016-6351-11e7-9943-d8df010f4a11.PNG />
<img width=320 src=https://user-images.githubusercontent.com/2364994/27968839-54f82662-6351-11e7-8bb9-f8ee315203fe.PNG />
2. That's how them look after restart (after 2-3 messages look fine)
<img width=320 src=https://user-images.githubusercontent.com/2364994/27968850-5a2390b8-6351-11e7-9516-517787e8adfe.PNG />
<img width=320 src=https://user-images.githubusercontent.com/2364994/27968845-57debf3a-6351-11e7-889b-c68bc8291fdd.PNG />
